platforms/allwinner-d1/core/src/drivers/sharp_display.rs

```diff
@@ -1,73 +1,393 @@
-use crate::drivers::spim::SpiSenderClient;
+//! Sharp display driver
+//!
+//! This is an early attempt at a "frame buffer" style display driver. It provides a
+//! [emb_display service][kernel::services::emb_display] server, and uses the
+//!
+//!
+//! This implementation is sort of a work in progress, it isn't really a *great*
+//! long-term solution, but rather "okay for now".
+//!
+//! A framebuffer of pixels is allocated for the entire display on registration.
+//! This could be, for example, 400x240 pixels.
+//!
+//! The driver will then allow for a certain number of "sub frames" to be requested.
+//!
+//! These sub frames could be for the entire display (400x240), or a portion of it,
+//! for example 200x120 pixels.
+//!
+//! Clients of the driver can draw into the sub-frames that they receive, then send
+//! them back to be rendered into the total frame. Any data in the client's sub-frame
+//! will replace the current contents of the whole frame buffer.
+
 use core::time::Duration;
-use kernel::{mnemos_alloc::containers::FixedVec, Kernel};
-
-pub async fn sharp_memory_display(k: &'static Kernel) {
-    k.sleep(Duration::from_millis(100)).await;
-    let mut spim = SpiSenderClient::from_registry(k).await.unwrap();
-
-    loop {
-        k.sleep(Duration::from_millis(100)).await;
-        let mut msg_1 = FixedVec::new(2).await;
-        msg_1.try_extend_from_slice(&[0x04, 0x00]).unwrap();
-        if spim.send_wait(msg_1).await.is_ok() {
-            break;
+
+use embedded_graphics::{
+    image::{Image, ImageRaw},
+    pixelcolor::Gray8,
+    prelude::*,
+    primitives::Rectangle,
+};
+use kernel::{
+    comms::kchannel::{KChannel, KConsumer},
+    maitake::sync::{Mutex, WaitCell},
+    mnemos_alloc::containers::{Arc, FixedVec},
+    registry::Message,
+    services::emb_display::{EmbDisplayService, FrameChunk, FrameError, Request, Response},
+    Kernel,
+};
+
+use crate::spim::SpiSenderClient;
+
+const WIDTH: usize = 400;
+const HEIGHT: usize = 240;
+const WIDTH_BYTES: usize = WIDTH / 8;
+
+/// Implements the [`EmbDisplayService`] service interface
+pub struct SharpDisplay;
+
+impl SharpDisplay {
+    /// Register the driver instance
+    ///
+    /// Registration will also start the simulated display, meaning that the display
+    /// window will appear.
+    pub async fn register(kernel: &'static Kernel, max_frames: usize) -> Result<(), FrameError> {
+        let frames = FixedVec::new(max_frames).await;
+        let mut linebuf = FixedVec::new((52 * 240) + 2).await;
+        for _ in 0..(52 * 240) + 2 {
```

Comment:
nit: should we use the constant here
```suggestion
        let mut linebuf = FixedVec::new((52 * HEIGHT) + 2).await;
        for _ in 0..(52 * HEIGHT) + 2 {
```